Source/DafnyCore/Pipeline/Compilation.cs

```diff
@@ -153,19 +172,28 @@ CompilationInput input
       var projectDirectory = Path.GetDirectoryName(projectPath)!;
       var filesMessage = string.Join("\n", result.Select(uri => Path.GetRelativePath(projectDirectory, uri.Uri.LocalPath)));
       if (filesMessage.Any()) {
-        errorReporter.Info(MessageSource.Parser, Project.StartingToken, "Files referenced by project are:" + Environment.NewLine + filesMessage);
-      } else {
-        errorReporter.Warning(MessageSource.Parser, CompilerErrors.ErrorId.None, Project.StartingToken, "Project references no files");
+        errorReporter.Info(MessageSource.Project, Project.StartingToken, "Files referenced by project are:" + Environment.NewLine + filesMessage);
       }
     }
 
-    return result;
+    if (!HasErrors && !result.Any()) {
+      errorReporter.Error(MessageSource.Project, CompilerErrors.ErrorId.None, Project.StartingToken,
+        "no Dafny source files were specified as input");
+    }
+
+    // Allow specifying the same file twice on the CLI
+    return result.DistinctBy(d => d.Uri).ToList();
   }
 
   private async Task<Program> ParseAsync() {
     try {
       await started.Task;
+      if (HasErrors) {
+        throw new OperationCanceledException();
+      }
```

Comment:
The name of this exception always bothered/confused me a little, and it feels worse using it in the CLI context now: it sounds like the user cancelled the pipeline, when it's really that the pipeline stoped early because of errors. 

Is this usage idiomatic for C# in general when hitting an "early return" as opposed to a parallel cancellation request?

Reply:
See above answer